### ISSUE TITLE
iunreg should be able to unregister a replica when the storage resource is down

### ISSUE THREAD

**Issue by tedgin** [irods/irods#6094]
This is a feature request, I think.

I'm running iRODS 4.2.8.

I have a data object with three good replicas. Two of the replicas are on the same storage resource with one of those being the special `replica/` folder in the top level of the vault. This storage resource has its status set to down. I want to remove the replica in `replica/`, so I try to use iunreg, but it refused to unregister the replica, reporting a `HIERARCHY_ERROR`.  If I change the status of the storage resource and its root to up, iunreg will unregister the replica. Since registration and unregistration are purely catalog database operations, the storage resource doesn't need to be up for iunreg to work. 

Would you all be willing to modify iunreg so that it doesn't require the storage resource to be up?

Here's an example.

```bash
ipc_admin@prod ~? ils -L /iplant/home/shared/project/fastq
  user                0 CyVerseRes;ds08Broker;ds08    640114282 2016-05-26.14:10 & fastq
    b15ba428cbb0306a590dbcf3d1c1d9be    generic    /vault/home/shared/prroject/fastq
  user                1 taccCorralRes;tacc    640114282 2017-03-03.19:51 & fastq
    b15ba428cbb0306a590dbcf3d1c1d9be    generic    /vault/replica/home/shared/project/fastq.3726226964
  user                2 taccCorralRes;tacc    640114282 2017-03-03.19:51 & fastq
    b15ba428cbb0306a590dbcf3d1c1d9be    generic    /vault/home/shared/project/fastq

ipc_admin@prod ~? iunreg -M -n1 /iplant/home/shared/project/fastq 
remote addresses: 206.207.252.46 ERROR: trimUtil: trim error for /iplant/home/shared/project/fastq.  status = -1803000 HIERARCHY_ERROR
Total size trimmed = 0.000 MB. Number of files trimmed = 0.

ipc_admin@prod ~? iadmin modresc tacc status up
ipc_admin@prod ~? iadmin modresc taccCorralRes status up
ipc_admin@prod ~? iunreg -M -n1 /iplant/home/shared/project/fastq 
Total size trimmed = 610.461 MB. Number of files trimmed = 1.
```

**Comment by trel** [comment]
related to / part of #4508

**Comment by alanking** [comment]
Some thoughts before a proposed resolution...

A resource being marked down is actually returned as an *error* in the voting logic. Strangely, the error code is not returned here.

Hierarchy resolution does consider 3 operations: create, open-for-write, and open-for-read. If we could add another operation for unreg and implement the voting logic in such a way that marked-down resources (as well as locality of reference and... most other considerations for voting) no longer results in a vote of 0, the system can honor the hierarchy resolution as it should and unreg can operate above the disk level as it should.

I don't think we should be dealing with this at the API level. Hierarchy resolution is intentionally opaque to enable flexibility in resource composition. A vote of 0 caused by the marked-down resource could just as well have been due to logical locking (inherent to any iRODS system) or a context string for a weighted passthru (configured policy). In a system with dozens or even hundreds of resources, considering the reason for the outcome of a particular hierarchy resolution could be cumbersome.

---

We have the following options for unregister completing as expected on a replica residing on a "marked down" resource:

1. Voting is ignored for unreg (bad - this is the current behavior and is captured in #5763)
2. Add unregister as a consideration in the hierarchy resolution (unreg votes, but logical locking gets exclusive rights to block the unreg) and remove "down" status, replacing it with hierarchy resolution (i.e. "voting") policy rather than catalog state. This would take care of this issue and (maybe) all of #4508.
3. An option can be added to the options considered in the passthru context string for unreg (captured in #6095 for discussion)

**Comment by kript** [comment]
> Some thoughts before a proposed resolution...
> 
> A resource being marked down is actually returned as an _error_ in the voting logic. Strangely, the error code is not returned here.

Perhaps it should be voting 0 instead? Or is that a naïve statement?

**Comment by alanking** [comment]
Even if the marked-down resource votes 0, there is no way for us to know the reason that a particular hierarchy votes 0. There is no way for the operation (in this case, unreg) to take action on the *reason* for the vote coming out the way it did because it does not (and currently cannot) know. This is actually the reason marked-down resources return an error instead of simply voting 0 - at some point, knowing this specific error was desired at the operation level.

If blocking all interactions with the resource via a passthru resource can produce the same effect, then the marked-down status is only useful from a monitoring perspective and would not need to be considered in voting if it even needs to exist at all.

**Comment by tedgin** [comment]
For purposes other than identifying which entry in the `ICAT.r_data_main` to delete, why does `iunreg` need to consider the resource at all?  Since `ireg` takes an existing, physical file and registers it in place into iRODS, it would make sense that `iunreg` is the inverse of `ireg`, i.e., `iunreg` unregisters a replica in iRODS, leaving the physical file where it is. If `iunreg` doesn't need to do anything with the physical file, it shouldn't matter whether or not the storage resource holding the file is accessible.  I can see this being important for `itrim`, which does need to delete the physical file.

**Comment by trel** [comment]
Yes, agreed @tedgin, it does not need access to the disk.   But the hierarchy still has to respond to the operation... which is why option 2 or 3 in https://github.com/irods/irods/issues/6094#issuecomment-1009181809 presented above are better than option 1.  The policy consistency is king - or nobody will be able to write policies that interact with one another in predictable ways.

I think option 2 is probably in our future ... where 'marked up' and 'marked down' are no longer how to manage these types of things.

**Comment by trel** [comment]
Can you set the resource status to 'up' - so hierarchy resolution can continue... but also set "read=0;write=0" in the resource's context string?

Would `iunreg` be able to function and remove the entries from the catalog?

**Comment by tedgin** [comment]
That doesn't work. `iunreg` requires access to the resource server.

In the following example I have stopped iRODS on 128.196.65.221. The storage resource `repl` is on that server, and it is a child of the passthru resource `replRes`.  The status is set to `up` on both `replRes` and `repl`. The context on `replRes` is set to `read=0.0;write=0.0`. The data object `README.md` has a replica on `replRes;repl`.

```console
rods@dev ~? ils -L README.md
  rods              0 ingestRes;ingest         2380 2024-03-25.14:43 & README.md
    0e39fd77535c9c382eb5d256e2ad71b3    generic    /mnt/ingest/home/rods/README.md
  rods              1 replRes;repl         2380 2024-03-25.14:43 & README.md
    0e39fd77535c9c382eb5d256e2ad71b3    generic    /mnt/repl/home/rods/README.md

rods@dev ~? iunreg -M -n 1 -N 1 README.md
remote addresses: 128.196.65.197 ERROR: trimUtil: trim error for /cyverse.dev/home/rods/README.md.  status = -305111 USER_SOCK_CONNECT_ERR, Connection refused
Total size trimmed = 0.000 MB. Number of files trimmed = 0.
Level 0: Specifying a minimum number of replicas to keep is deprecated.

rods@dev ~? iunreg -M -N 1 -S replRes README.md
remote addresses: 128.196.65.197 ERROR: trimUtil: trim error for /cyverse.dev/home/rods/README.md.  status = -305111 USER_SOCK_CONNECT_ERR, Connection refused
Total size trimmed = 0.000 MB. Number of files trimmed = 0.
Level 0: Specifying a minimum number of replicas to keep is deprecated.
```

**Comment by tedgin** [comment]
Ok. This isn't very intuitive. I got it to work by providing the storage resource name to `iunreg` instead of the root resource name or the replica number.

```console
rods@dev ~? iunreg -M -N 1 -S repl README.md
Total size trimmed = 0.001 MB. Number of files trimmed = 1.

rods@dev ~? ils -l README.md
  rods              0 ingestRes;ingest         2380 2024-03-25.14:43 & README.md
```

I found my workaround! Restating it, status needs to be set to `up` on both the storage resource and its parent. context needs to be set to `read=0.0;write=0.0` on the parent. When calling `iunreg`, use the `-S` option to specify the name of the storage resource.

**Comment by trel** [comment]
By using the resource name, you’re bypassing the voting mechanism.

**Comment by tedgin** [comment]
Shouldn't resource number (`-n` option) also by bypass voting?

**Comment by trel** [comment]
I would expect so, yes.      Perhaps @alanking has some insight here regarding the difference.

**Comment by alanking** [comment]
`-n` does not bypass voting; it denotes a "requested replica number". If a replica is requested by replica number, resources which host replicas which do not have that replica number vote "low" (0.25) and the resource which hosts the replica with the requested replica number votes "high" (1.00). Once all of the votes have been collected, if the winning vote is not the replica with the requested replica number, it results in a failure to resolve the resource hierarchy.

Here's the documentation describing voting behavior (case 4 for unlink is the relevant bit for this discussion, I believe):
https://docs.irods.org/4.3.3/plugins/composable_resources/#unixfilesystem-resource-voting

Here's the code for voting at the storage resource level:
https://github.com/irods/irods/blob/61ab2379c2d92a3d47a6ee3f6f59bc79ed6c32b0/server/core/src/voting.cpp#L81-L84

Here's the relevant code for deciding the winning vote:
https://github.com/irods/irods/blob/61ab2379c2d92a3d47a6ee3f6f59bc79ed6c32b0/server/core/src/irods_resource_redirect.cpp#L248-L273

While I'm glad you got things working, I would also argue that the workaround being used to bypass voting via resource name is a bug in `iunreg`. The resource name should only be targeting a root resource, not a leaf resource. We also allow use of the `RESC_HIER_STR_KW` to completely bypass voting in all of our operations. It is my opinion that neither of these should be allowed, but it is necessary for now. Hopefully the workaround won't be necessary once we massage our unregistering mechanisms to not require access to the vault (because it really doesn't need it).
